### PR TITLE
Http client

### PR DESCRIPTION
**Pull request by shmish111**
There are 2 parts to this PR (in 2 commits)

1. Clean up the HTTP server so that it reuses the existing machinery from `Trace`
2. Create an HTTP client for all the endpoints

I am actually unsure where to go from here. The client can't fit into `Trace` since that contains `EmulatedWalletApi` which is a wrapper around `StateT WalletState`. Basically I'm unsure how or where we now use this client?

Anyway, that would be the remit of another PR, I don't think it should stop this PR being merged.

### CODE REVIEW

File: wallet-api/src/Wallet/Emulator/Types.hs [hunk 1]
```diff
@@ -200,7 +200,7 @@ isValidated txn = do
 data Event n a where
     -- | An direct action performed by a wallet. Usually represents a "user action", as it is
     -- triggered externally.
-    WalletAction :: Wallet -> n () -> Event n [Tx]
+    WalletAction :: Wallet -> n b -> Event n ([Tx], Either WalletAPIError b)
```
Comment (michaelpj) [inline wallet-api/src/Wallet/Emulator/Types.hs]: Hm, no, the signature is deliberately restricted here. WalletAPI errors don't affect the trace other than aborting the wallet action, which I'm pretty sure they'll already do. Why did you change this?

**Comment by michaelpj** [comment]
I don't understand your question:

> I am actually unsure where to go from here. The client can't fit into Trace since that contains EmulatedWalletApi which is a wrapper around StateT WalletState. Basically I'm unsure how or where we now use this client?

My initial thought was that we would do something like this:
- `process trace` produces a `MonadEmulator m => m a`.
- We provide functions to lift such a value into one of:
    - The in-memory interpretation provided by the main emulator module.
    - The implementation used by the HTTP backend.
    - A HTTP client implementation which makes calls to the backend.

That obviously doesn't work since `MonadEmulator` expects to operate in terms of the underlying state. So we want something that interprets the trace directly into a series of calls to the backend. The idea is still that you run exactly the same trace in any of the settings.

File: wallet-api/src/Wallet/Emulator/Types.hs (hunk 1 above)
Comment (michaelpj) [reply wallet-api/src/Wallet/Emulator/Types.hs]: ... and wallet actions deliberately don't get to produce a value other than unit.

Comment (shmish111) [reply wallet-api/src/Wallet/Emulator/Types.hs]: In order for the HTTP server to return the correct thing using the Trace api, it will need to know the result of the wallet action. Why does `WalletAction` not return the type of the wallet action currently, it seems strange to throw away the result of the action? I could combine `WalletAPIError` and `AssertionError`? It seems strange that the error type of `MonadEmulator` is an `AssertionError` anyway. That way we would have `WalletAction :: Wallet -> n b -> Event n ([Tx], b)`.

Alternatively I can not use the Trace api for the http server?

Comment (shmish111) [reply wallet-api/src/Wallet/Emulator/Types.hs]: > and wallet actions deliberately don't get to produce a value other than unit.

but they do in the `WalletAPI`, why don't they here?

**Comment by michaelpj** [comment]
Okay, I think there's some confusion about what's going on here.

`WalletAPI`:
The wallet API is the API to the wallet. This includes actions that users can take with their wallets, and should return them errors etc. This is the thing that the real wallet will eventually implement, and which we're hoping to influence the design of. Most of the wallet endpoints in the HTTP server should be implementing this.

`Trace`:
This represents a series of events from the perspective of the smart contract system. That means it has a slightly different view: in particular, if wallet actions fail they just fail and that's the wallet's problem. The trace will continue just fine (as the real world would happily chug along if you submit a bad wallet action).

The only effect that wallet actions have in the `Trace` world is the side effect of adding transactions to the pending transaction pool. That's why their entry in the `Event` GADT has the signature that it does - you give us some action which produces no value (we could say it produces a value but we ignore it, I guess), and we get out the transactions that it produces (possibly none if it fails!).

Evaluating the `WalletAction` `Event`s in a `Trace` therefore means evaluating the wallet action in your context, and then throwing away the error code and result and just getting the transactions.

---

Writing that has clarified my thinking too: we should *not* be evaluating `Trace`s or `Event`s in the HTTP backend, we should be evaluating *wallet API actions*. The emulator-specific endpoints are different, of course, since they deal with things that aren't in the wallet api, like assertions. But still, we shouldn't try to make the backend evaluate a `Trace` or even an `Event` (my bad for suggesting that we should!).

What *does* need to evaluate a `Trace` is HTTP-client based evaluator that will evaluate wallet actions using the HTTP wallet API, and then throw away the result value/error code and just return the new transactions.

--

Does that help?

**Comment by michaelpj** [comment]
Concretely, I think this means we should have an implementation of `WalletApi` in terms of the HTTP client. We could concievably then have a `RestOfTraceAPI` class (which would do things like validate transactions, run assertions etc.), implement that, and then write a generic `Trace` evaluator with both as a constraint.

**Comment by shmish111** [comment]
OK, that helps, it's exactly what I thought _before_ I started writing the code :-)

I was trying to re-use as much as possible but I think this PR has shown that that doesn't really work. I can refactor `Types.hs` a little bit in order to make some of the tools available but not use the trace stuff directly for the server.

Also it's easy to split up the API, it's already done actually, just need to make 2 http clients instead of 1 and then wrap up the wallet client behind the wallet API (my original intention).

So for this PR, I will put back the changes to the Trace stuff and change the HTTP server to not use it, I will also create a `WalletAPI` implementation for the http client and then perhaps see where we are. That will probably take me the rest of today anyway.

**Comment by michaelpj** [comment]
Cool, sorry for the mixed messages! I don't have this stuff quite loaded in my head at the moment.

File: plutus-use-cases/test/Spec/Crowdfunding.hs [hunk 2]
```diff
@@ -59,7 +59,9 @@ contrib3 = contrib ds (Wallet 3) where
 
 -- | Collect the contributions of a crowdfunding campaign
 collect :: Wallet -> CampaignPLC -> [(TxOutRef', Wallet, UTXO.Value)] -> Trace [Tx]
-collect w c contributions = walletAction w $ CF.collect c ins where
+collect w c contributions =
+  fmap fst . walletAction w $ CF.collect c ins
```
Comment (vmchale) [inline plutus-use-cases/test/Spec/Crowdfunding.hs]: Easier to read as `fst <$> walletAction w $ ...` if precedence works with that.

**Comment by michaelpj** [comment]
One further thing that occurs to me: what we currently think of as `WalletAPI` is actually a combination of `WalletUserAPI` and `WalletControlAPI`. The user should not be able to tell the wallet that it's received a notification - that's controlled by the blockchain client it's using!

So we should probably break this up even more.

File: stack.yaml [hunk 3]
```diff
@@ -21,7 +21,7 @@ packages:
   extra-dep: true
 
 extra-deps:
-- cryptonite-0.22
+- cryptonite-0.25
```
Comment (vmchale) [inline stack.yaml]: Is this the version they're using upstream? If this needs a specific version of above, it would be advisable to add a lower version bound to the `.cabal` file.

File: wallet-api/src/Wallet/Emulator/Http.hs [hunk 4]
```diff
@@ -192,13 +167,46 @@ walletHandlers state = hoistServer api (runM state) $ walletApi :<|> controlApi
...
       blockHeight
+    assertionsApi = assertOwnFundsEq :<|> assertIsValidated
+
+assertOwnFundsEq ::
+     (MonadError ServantErr m, MonadReader State m, MonadIO m)
+  => Wallet
+  -> Value
+  -> m NoContent
+assertOwnFundsEq wallet value =
+  fmap (const NoContent) $ runTrace $ Types.assertOwnFundsEq wallet value
```
Comment (vmchale) [inline wallet-api/src/Wallet/Emulator/Http.hs]: `fmap (const x) y` is just `(x <$ y)` or `(y <$ x)`

**Comment by j-mueller** [comment]
> The user should not be able to tell the wallet that it's received a notification 

It's `walletRecvNotifications :: Wallet -> [Notification] -> Trace [Tx]` so it requires a `MonadEmulator` constraint to run - I think the `WalletAPI` class can be implemented by a real wallet as it is (at least that was the intention behind splitting up the API into two parts in (CGP-377)[https://iohk.myjetbrains.com/youtrack/issue/CGP-377])

**Comment by j-mueller** [comment]
We should probably have a call or something to agree on a definition because I have been working with different assumptions about `Trace`. In particular

> 
> `WalletAPI`:
> The wallet API is the API to the wallet. (...)

Agree 

> 
> `Trace`:
> This represents a series of events from the perspective of the smart contract system. (...)

Currently it is a series of events from the perspective of the emulator though. We can manipulate the global blockchain/mockchain state in a `Trace`, for example we we can say when a new block is created, and when wallets are notified of that.

> 
> The only effect that wallet actions have in the `Trace` world is the side effect of adding transactions to the pending transaction pool. (...)

Yes

> Writing that has clarified my thinking too: we should _not_ be evaluating `Trace`s or `Event`s in the HTTP backend, we should be evaluating _wallet API actions_. The emulator-specific endpoints are different, of course, since they deal with things that aren't in the wallet api, like assertions. But still, we shouldn't try to make the backend evaluate a `Trace` or even an `Event` (my bad for suggesting that we should!).

The intention behind this task (standalone mockchain) was to have an executable that implements the wallet HTTP API so that we can run plutus coordination code (compiled to JS) against it without having to rely on the real blockchain. But to achieve that the executable also needs to emulate the blockchain in order to give realistic responses for the wallet API. And the emulator in turn is controlled by `Trace` actions. If the standalone mockchain only implements the wallet API, but not the emulator API, then how do we advance the state of the blockchain?

So I think we should try to make the backend evaluate both `WalletAPI` and `Trace` actions, and that's what the two parts of the API in [CGP-377](https://iohk.myjetbrains.com/youtrack/issue/CGP-377) are about. If that's wrong then we need to change the task on YT. But again I think we should arrange a call to clarify this.

**Comment by michaelpj** [comment]
> I think the WalletAPI class can be implemented by a real wallet as it is 

Yes, but the *real* wallet shouldn't expose `walletRecvNotification` to a user! We want to segment those bits of the API.

> Currently it is a series of events from the perspective of the emulator though. We can manipulate the global blockchain/mockchain state in a Trace, for example we we can say when a new block is created, and when wallets are notified of that.

Sure, that's more-or-less what I meant.

> If the standalone mockchain only implements the wallet API, but not the emulator API, then how do we advance the state of the blockchain?

I definitely did not mean that it should only implement the wallet API, just that the bit which corresponds (will correspond) to the real wallet API should be clearly separate.

I agree having a call would be useful.

**Comment by j-mueller** [comment]
After talking with @michaelpj I think we can summarise the capabilities of the "standalone mockchain" as follows:

1) `WalletAPI`: Anything that can be run against a real wallet. Corresponds to 
    * Part I in https://iohk.myjetbrains.com/youtrack/issue/CGP-377
	* `Wallet.Emulator.Http.WalletAPI`
2) Mockchain control API. Anything that controls the emulated blockchain. Corresponds to
    * Part II in https://iohk.myjetbrains.com/youtrack/issue/CGP-377 *except* `walletRecvNotifications` (see (3.) below)
	* `Wallet.Emulator.Http.ControlAPI` *except* the two `"wallets" :> Capture "walletid" Wallet :> "notifications"` endpoints
3) Wallet control API **(new)**. The ability to notify a wallet of blockchain events. Represents communication from the cardano node to the wallet. This should be treated separately (perhaps a class `NotifyWallet m`) because in the future this might also be implemented by the real wallet for testing purposes (for example with a debug flag). This corresponds to
    * `"emulator" :> "wallets" :> Capture "walletid" Wallet :> "notifications" :> "block-validation" :> ReqBody '[ JSON] Block :> Post '[ JSON] ()` and
	* `"emulator" :> "wallets" :> Capture "walletid" Wallet :> "notifications" :> "block-height" :> ReqBody '[ JSON]`
	* (the `emulator` bit should be replaced by `wallet-control` or similar)

The confusion was mostly about `walletRecvNotifications` which currently resides in the control API but should become its own API. 

So for the PR this means:

* https://github.com/input-output-hk/plutus/pull/214/files#diff-da394a94497e501c3cc7a1f43f3d9a81R66 should be `type API = WalletAPI :<|> ControlAPI :<|> WalletControlAPI` with the appropriate types (ie. merge `AssertionsAPI` into `ControlAPI` and extract the `blockValidated` and `blockHeight` endpoints into `WalletControlAPI`)
* Revert the changes to the `WalletAction` constructor https://github.com/input-output-hk/plutus/pull/214/files#diff-66d9ee4fe6c0f85f4af266395458289bR203 - we care about the return values only if we use `WalletAPI` directly, not inside a `Trace`
* Another issue that should be changed (unless it adds duplication) is that every wallet action is wrapped in a trace in `runWalletAction`. Example: https://github.com/input-output-hk/plutus/pull/214/files#diff-da394a94497e501c3cc7a1f43f3d9a81R112 Maybe it is possible to implement this in a way that doesn't require `Trace`.

In the future we should add an implementation of `WalletAPI` that uses a subset of the functions from https://github.com/input-output-hk/plutus/pull/214/files#diff-b7bde62939a8bbbc0fcb561c2975656aR4 and returns the correct results.

File: plutus-use-cases/test/Spec/Crowdfunding.hs (hunk 2 above)
Comment (shmish111) [reply plutus-use-cases/test/Spec/Crowdfunding.hs]: yeah precedence meant it didn't work. reverting anyway so it doesn't matter

File: wallet-api/src/Wallet/Emulator/Http.hs (hunk 4 above)
Comment (shmish111) [reply wallet-api/src/Wallet/Emulator/Http.hs]: nice

File: stack.yaml (hunk 3 above)
Comment (shmish111) [reply stack.yaml]: done

**Comment by shmish111** [comment]
OK, I've made the changes suggested I think. I haven't wrapped up the client yet as I'm not quite sure how the classes will work but I'd prefer to keep that for the next PR as this is getting quite big. This PR should just now represent a refactoring of how the http server works as well as creating a client.

I will do a squash once it is approved.

In reality it was more of a learning experience for everyone :-)

**Comment by j-mueller** [comment]
Looks good now, I think you have to run `generate.sh` to update the nix expressions.

**Comment by michaelpj** [comment]
hydra is being silly, but this is failing the stylish haskell check: https://hydra.iohk.io/build/296337

**Comment by shmish111** [comment]
oops, thought I'd run fixStylishHaskell already. pushed now.

**Comment by michaelpj** [comment]
Okay, this has in fact gone green.

File: stack.yaml (hunk 3 above)
Comment (vmchale) [reply stack.yaml]: Wait, so does this agree with upstream cardano libraries?

Comment (shmish111) [reply stack.yaml]: yes: https://github.com/input-output-hk/cardano-sl/blob/8e26a7ac346d84135a4e178af58c144ff80ea222/cabal.project.freeze#L44
